File: apk+ublock/app/src/main/java/com/ubour/vpn/core/SingboxManager.kt
package com.ubour.vpn.core

import android.content.Context
import android.util.Log
import com.ubour.vpn.warp.WarpConfig
import org.json.JSONArray
import org.json.JSONObject
import java.io.File

object SingboxManager {
    private const val TAG = "SingboxManager"
    private var process: Process? = null
    const val SOCKS_PORT = 10809

    fun isRunning(): Boolean {
        return process?.isAlive == true
    }

    fun startWarp(context: Context, warpConfig: WarpConfig, enableAdBlock: Boolean = true): Boolean {
        stop()
        val configFile = generateWarpConfig(context, warpConfig, enableAdBlock)
        return startProcess(context, configFile)
    }

    fun startAdBlockOnly(context: Context): Boolean {
        stop()
        val configFile = generateAdBlockOnlyConfig(context)
        return startProcess(context, configFile)
    }

    fun startVpnAndAdBlock(context: Context, byedpiPort: Int = 1080): Boolean {
        stop()
        val configFile = generateVpnAndAdBlockConfig(context, byedpiPort)
        return startProcess(context, configFile)
    }

    fun startVless(context: Context, vlessUrl: String, enableAdBlock: Boolean = true): Boolean {
        stop()
        val configFile = generateVlessConfig(context, vlessUrl, enableAdBlock) ?: return false
        return startProcess(context, configFile)
    }

    fun stop() {
        try {
            process?.destroy()
            process = null
            Log.i(TAG, "Sing-box process stopped")
        } catch (e: Exception) {
            Log.e(TAG, "Error stopping Sing-box: ${e.message}")
        }
    }

    fun getExecutableBinary(context: Context): File? {
        val nativeDir = context.applicationInfo.nativeLibraryDir
        val nativeBinary = File(nativeDir, "libsingbox.so")
        if (nativeBinary.exists()) {
            if (!nativeBinary.canExecute()) {
                nativeBinary.setExecutable(true, true)
            }
            return nativeBinary
        }
        val fallbackBinary = File(context.filesDir, "libsingbox.so")
        if (fallbackBinary.exists()) {
            if (!fallbackBinary.canExecute()) {
                fallbackBinary.setExecutable(true, true)
            }
            return fallbackBinary
        }
        Log.e(TAG, "libsingbox.so not found in nativeLibraryDir: $nativeDir")
        return null
    }

    private fun startProcess(context: Context, configFile: File): Boolean {
        val binaryToRun = getExecutableBinary(context)
        if (binaryToRun == null || !binaryToRun.exists()) {
            Log.e(TAG, "No valid singbox binary found in nativeLibraryDir!")
            return false
        }

        return try {
            val pb = ProcessBuilder(
                binaryToRun.absolutePath,
                "run",
                "-c",
                configFile.absolutePath
            ).directory(context.filesDir)
             .redirectErrorStream(true)

            process = pb.start()

            // Start log consumer thread
            Thread {
                try {
                    val reader = process?.inputStream?.bufferedReader()
                    while (isRunning()) {
                        val line = reader?.readLine() ?: break
                        Log.i("SingboxCore", line)
                    }
                } catch (_: Exception) {}
            }.start()

            Thread.sleep(400)
            val alive = isRunning()
            Log.i(TAG, "Sing-box process status: isAlive=$alive (Path: ${binaryToRun.absolutePath})")
            alive
        } catch (e: Exception) {
            Log.e(TAG, "Failed to start Sing-box process: ${e.message}", e)
            false
        }
    }

    private fun generateWarpConfig(context: Context, warp: WarpConfig, enableAdBlock: Boolean = true): File {
        val addrs = JSONArray().apply {
            put(warp.localIpv4)
            if (!warp.localIpv6.isNullOrBlank()) {
                put(warp.localIpv6)
            }
        }

        val peerObj = JSONObject().apply {
            put("server", warp.endpointHost)
            put("server_port", warp.endpointPort)
            put("public_key", warp.peerPublicKey)
            put("allowed_ips", JSONArray().apply {
                put("0.0.0.0/0")
                put("::/0")
            })
            if (warp.reserved.size == 3) {
                put("reserved", JSONArray().apply {
                    warp.reserved.forEach { put(it) }
                })
            }
        }

        val warpOutbound = JSONObject().apply {
            put("type", "wireguard")
            put("tag", "warp-out")
            put("server", warp.endpointHost)
            put("server_port", warp.endpointPort)
            put("local_address", addrs)
            put("private_key", warp.privateKey)
            put("peer_public_key", warp.peerPublicKey)
            if (warp.reserved.size == 3) {
                put("reserved", JSONArray().apply {
                    warp.reserved.forEach { put(it) }
                })
            }
            put("peers", JSONArray().apply {
                put(peerObj)
            })
            put("mtu", 1280)
        }

        val dnsFilterOutbound = JSONObject().apply {
            put("type", "direct")
            put("tag", "dns-filter-out")
            put("override_address", "127.0.0.1")
            put("override_port", 5353)
        }

        val root = JSONObject().apply {
            put("log", JSONObject().apply {
                put("level", "info")
            })
            put("inbounds", JSONArray().apply {
                put(JSONObject().apply {
                    put("type", "socks")
                    put("tag", "socks-in")
                    put("listen", "127.0.0.1")
                    put("listen_port", SOCKS_PORT)
                })
            })
            put("outbounds", JSONArray().apply {
                put(warpOutbound)
                if (enableAdBlock) {
                    put(dnsFilterOutbound)
                }
                put(JSONObject().apply {
                    put("type", "direct")
                    put("tag", "direct")
                })
            })
            put("route", JSONObject().apply {
                put("rules", JSONArray().apply {
                    if (enableAdBlock) {
                        put(JSONObject().apply {
                            put("port", JSONArray().apply { put(53) })
                            put("outbound", "dns-filter-out")
                        })
                    }
                    put(JSONObject().apply {
                        put("inbound", JSONArray().apply { put("socks-in") })
                        put("outbound", "warp-out")
                    })
                })
                put("final", "warp-out")
            })
        }

        val configFile = File(context.filesDir, "singbox_warp.json")
        configFile.writeText(root.toString(2))
        return configFile
    }

    private fun generateAdBlockOnlyConfig(context: Context): File {
        val dnsFilterOutbound = JSONObject().apply {
            put("type", "direct")
            put("tag", "dns-filter-out")
            put("override_address", "127.0.0.1")
            put("override_port", 5353)
        }

        val root = JSONObject().apply {
            put("log", JSONObject().apply {
                put("level", "info")
            })
            put("inbounds", JSONArray().apply {
                put(JSONObject().apply {
                    put("type", "socks")
                    put("tag", "socks-in")
                    put("listen", "127.0.0.1")
                    put("listen_port", SOCKS_PORT)
                })
            })
            put("outbounds", JSONArray().apply {
                put(JSONObject().apply {
                    put("type", "direct")
                    put("tag", "direct")
                })
                put(dnsFilterOutbound)
            })
            put("route", JSONObject().apply {
                put("rules", JSONArray().apply {
                    put(JSONObject().apply {
                        put("port", JSONArray().apply { put(53) })
                        put("outbound", "dns-filter-out")
                    })
                    put(JSONObject().apply {
                        put("inbound", JSONArray().apply { put("socks-in") })
                        put("outbound", "direct")
                    })
                })
                put("final", "direct")
            })
        }

        val configFile = File(context.filesDir, "singbox_adblock.json")
        configFile.writeText(root.toString(2))
        return configFile
    }

    private fun generateVpnAndAdBlockConfig(context: Context, byedpiPort: Int): File {
        val byedpiOutbound = JSONObject().apply {
            put("type", "socks")
            put("tag", "byedpi-out")
            put("server", "127.0.0.1")
            put("server_port", byedpiPort)
        }

        val dnsFilterOutbound = JSONObject().apply {
            put("type", "direct")
            put("tag", "dns-filter-out")
            put("override_address", "127.0.0.1")
            put("override_port", 5353)
        }

        val root = JSONObject().apply {
            put("log", JSONObject().apply {
                put("level", "info")
            })
            put("inbounds", JSONArray().apply {
                put(JSONObject().apply {
                    put("type", "socks")
                    put("tag", "socks-in")
                    put("listen", "127.0.0.1")
                    put("listen_port", SOCKS_PORT)
                })
            })
            put("outbounds", JSONArray().apply {
                put(byedpiOutbound)
                put(dnsFilterOutbound)
                put(JSONObject().apply {
                    put("type", "direct")
                    put("tag", "direct")
                })
            })
            put("route", JSONObject().apply {
                put("rules", JSONArray().apply {
                    put(JSONObject().apply {
                        put("port", JSONArray().apply { put(53) })
                        put("outbound", "dns-filter-out")
                    })
                    put(JSONObject().apply {
                        put("inbound", JSONArray().apply { put("socks-in") })
                        put("outbound", "byedpi-out")
                    })
                })
                put("final", "byedpi-out")
            })
        }

        val configFile = File(context.filesDir, "singbox_vpn_adblock.json")
        configFile.writeText(root.toString(2))
        return configFile
    }

    private fun generateVlessConfig(context: Context, vlessUrl: String, enableAdBlock: Boolean = true): File? {
        try {
            if (!vlessUrl.startsWith("vless://", ignoreCase = true)) return null
            val raw = vlessUrl.substring(8)
            val atIdx = raw.indexOf('@')
            val colonIdx = raw.indexOf(':', atIdx)
            val qIdx = raw.indexOf('?', colonIdx)

            val uuid = raw.substring(0, atIdx)
            val host = raw.substring(atIdx + 1, colonIdx)
            val portStr = if (qIdx != -1) raw.substring(colonIdx + 1, qIdx) else raw.substring(colonIdx + 1)
            val port = portStr.toIntOrNull() ?: 443

            var sni = host
            var pbk = ""
            var sid = ""
            var fp = "chrome"

            if (qIdx != -1) {
                val query = raw.substring(qIdx + 1)
                val params = query.split("&")
                for (p in params) {
                    val kv = p.split("=")
                    if (kv.size == 2) {
                        when (kv[0].lowercase()) {
                            "sni" -> sni = kv[1]
                            "pbk" -> pbk = kv[1]
                            "sid" -> sid = kv[1]
                            "fp" -> fp = kv[1]
                        }
                    }
                }
            }

            val dnsFilterOutbound = JSONObject().apply {
                put("type", "direct")
                put("tag", "dns-filter-out")
                put("override_address", "127.0.0.1")
                put("override_port", 5353)
            }

            val root = JSONObject().apply {
                put("log", JSONObject().put("level", "info"))
                put("inbounds", JSONArray().apply {
                    put(JSONObject().apply {
                        put("type", "socks")
                        put("tag", "socks-in")
                        put("listen", "127.0.0.1")
                        put("listen_port", SOCKS_PORT)
                    })
                })
                put("outbounds", JSONArray().apply {
                    put(JSONObject().apply {
                        put("type", "vless")
                        put("tag", "vless-out")
                        put("server", host)
                        put("server_port", port)
                        put("uuid", uuid)
                        put("flow", "xtls-rprx-vision")
                        put("tls", JSONObject().apply {
                            put("enabled", true)
                            put("server_name", sni)
                            put("utls", JSONObject().apply {
                                put("enabled", true)
                                put("fingerprint", fp)
                            })
                            if (pbk.isNotBlank()) {
                                put("reality", JSONObject().apply {
                                    put("enabled", true)
                                    put("public_key", pbk)
                                    put("short_id", sid)
                                })
                            }
                        })
                    })
                    if (enableAdBlock) {
                        put(dnsFilterOutbound)
                    }
                })
                put("route", JSONObject().apply {
                    put("rules", JSONArray().apply {
                        if (enableAdBlock) {
                            put(JSONObject().apply {
                                put("port", JSONArray().apply { put(53) })
                                put("outbound", "dns-filter-out")
                            })
                        }
                        put(JSONObject().apply {
                            put("inbound", JSONArray().apply { put("socks-in") })
                            put("outbound", "vless-out")
                        })
                    })
                    put("final", "vless-out")
                })
            }

            val configFile = File(context.filesDir, "singbox_vless.json")
            configFile.writeText(root.toString(2))
            return configFile
        } catch (e: Exception) {
            Log.e(TAG, "Failed to parse vless URL: ${e.message}")
            return null
        }
    }
}

File: windows/src/Ubour/Services/SingboxManager.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.IO;
using System.Net;
using System.Net.Sockets;
using System.Text.Json;
using System.Threading;
using Ubour.Models;

namespace Ubour.Services;

public class SingboxManager
{
    private Process? _process;
    private string? _currentConfigFile;
    public bool IsRunning => _process != null && !_process.HasExited;

    public bool StartWarp(string baseDir, WarpConfig warp, bool enableAdBlock = true)
    {
        Stop();
        string configPath = GenerateWarpConfig(warp, enableAdBlock);
        bool started = StartProcess(baseDir, configPath);
        if (started)
        {
            ProxyManager.EnableProxy("127.0.0.1:2080");
        }
        return started;
    }

    public bool StartAdBlockOnly(string baseDir)
    {
        Stop();
        string configPath = GenerateAdBlockOnlyConfig();
        bool started = StartProcess(baseDir, configPath);
        if (started)
        {
            ProxyManager.EnableProxy("127.0.0.1:2080");
        }
        return started;
    }

    public bool StartVless(string baseDir, string vlessUrl, bool enableAdBlock = true)
    {
        Stop();
        string? configPath = GenerateVlessConfig(vlessUrl, enableAdBlock);
        if (string.IsNullOrEmpty(configPath)) return false;
        bool started = StartProcess(baseDir, configPath);
        if (started)
        {
            ProxyManager.EnableProxy("127.0.0.1:2080");
        }
        return started;
    }

    public bool VerifyRunning(int port = 2080, int timeoutMs = 2500)
    {
        if (_process == null || _process.HasExited) return false;
        try
        {
            using var client = new TcpClient();
            var task = client.ConnectAsync(IPAddress.Loopback, port);
            if (!task.Wait(timeoutMs)) return false;
            return client.Connected;
        }
        catch
        {
            return false;
        }
    }

    private string GenerateWarpConfig(WarpConfig warp, bool enableAdBlock)
    {
        string tmpDir = Path.Combine(Path.GetTempPath(), "Ubour");
        Directory.CreateDirectory(tmpDir);
        string configFile = Path.Combine(tmpDir, "singbox_warp.json");

        var dnsObj = BuildDnsConfig(enableAdBlock);
        var routeRules = BuildRouteRules(enableAdBlock, "warp-ep");

        var configObj = new
        {
            log = new { level = "info" },
            dns = dnsObj,
            endpoints = new object[]
            {
                new
                {
                    type = "wireguard",
                    tag = "warp-ep",
                    system = false,
                    name = "warp",
                    address = new string[] { warp.LocalIpv4, warp.LocalIpv6 },
                    private_key = warp.PrivateKey,
                    mtu = 1280,
                    peers = new object[]
                    {
                        new
                        {
                            address = warp.EndpointHost,
                            port = warp.EndpointPort,
                            public_key = warp.PeerPublicKey,
                            allowed_ips = new string[] { "0.0.0.0/0", "::/0" },
                            reserved = new int[] { warp.Reserved[0], warp.Reserved[1], warp.Reserved[2] }
                        }
                    }
                }
            },
            inbounds = new object[]
            {
                new
                {
                    type = "mixed",
                    tag = "mixed-in",
                    listen = "127.0.0.1",
                    listen_port = 2080
                }
            },
            outbounds = new object[]
            {
                new
                {
                    type = "direct",
                    tag = "direct"
                }
            },
            route = new
            {
                rules = routeRules,
                auto_detect_interface = true
            }
        };

        File.WriteAllText(configFile, JsonSerializer.Serialize(configObj, new JsonSerializerOptions { WriteIndented = true }));
        _currentConfigFile = configFile;
        return configFile;
    }

    private string GenerateAdBlockOnlyConfig()
    {
        string tmpDir = Path.Combine(Path.GetTempPath(), "Ubour");
        Directory.CreateDirectory(tmpDir);
        string configFile = Path.Combine(tmpDir, "singbox_adblock.json");

        var dnsObj = BuildDnsConfig();
        var routeRules = BuildRouteRules(true, "direct");

        var configObj = new
        {
            log = new { level = "info" },
            dns = dnsObj,
            inbounds = new object[]
            {
                new
                {
                    type = "mixed",
                    tag = "mixed-in",
                    listen = "127.0.0.1",
                    listen_port = 2080
                }
            },
            outbounds = new object[]
            {
                new
                {
                    type = "direct",
                    tag = "direct"
                }
            },
            route = new
            {
                rules = routeRules,
                auto_detect_interface = true
            }
        };

        File.WriteAllText(configFile, JsonSerializer.Serialize(configObj, new JsonSerializerOptions { WriteIndented = true }));
        _currentConfigFile = configFile;
        return configFile;
    }

    private string? GenerateVlessConfig(string vlessUrl, bool enableAdBlock)
    {
        try
        {
            if (!vlessUrl.StartsWith("vless://", StringComparison.OrdinalIgnoreCase)) return null;
            string raw = vlessUrl.Substring(8);
            int atIdx = raw.IndexOf('@');
            int colonIdx = raw.IndexOf(':', atIdx);
            int qIdx = raw.IndexOf('?', colonIdx);

            string uuid = raw.Substring(0, atIdx);
            string host = raw.Substring(atIdx + 1, colonIdx - atIdx - 1);
            string portStr = qIdx != -1 ? raw.Substring(colonIdx + 1, qIdx - colonIdx - 1) : raw.Substring(colonIdx + 1);
            int port = int.TryParse(portStr, out int p) ? p : 443;

            string sni = host;
            string pbk = "";
            string sid = "";

            if (qIdx != -1)
            {
                string query = raw.Substring(qIdx + 1);
                string[] parts = query.Split('&');
                foreach (var part in parts)
                {
                    string[] kv = part.Split('=');
                    if (kv.Length == 2)
                    {
                        if (kv[0].Equals("sni", StringComparison.OrdinalIgnoreCase)) sni = kv[1];
                        if (kv[0].Equals("pbk", StringComparison.OrdinalIgnoreCase)) pbk = kv[1];
                        if (kv[0].Equals("sid", StringComparison.OrdinalIgnoreCase)) sid = kv[1];
                    }
                }
            }

            string tmpDir = Path.Combine(Path.GetTempPath(), "Ubour");
            Directory.CreateDirectory(tmpDir);
            string configFile = Path.Combine(tmpDir, "singbox_vless.json");

            var dnsObj = BuildDnsConfig(enableAdBlock);
            var routeRules = BuildRouteRules(enableAdBlock, "vless-out");

            var configObj = new
            {
                log = new { level = "info" },
                dns = dnsObj,
                inbounds = new object[]
                {
                    new
                    {
                        type = "mixed",
                        tag = "mixed-in",
                        listen = "127.0.0.1",
                        listen_port = 2080
                    }
                },
                outbounds = new object[]
                {
                    new
                    {
                        type = "vless",
                        tag = "vless-out",
                        server = host,
                        server_port = port,
                        uuid = uuid,
                        flow = "xtls-rprx-vision",
                        tls = new
                        {
                            enabled = true,
                            server_name = sni,
                            utls = new
                            {
                                enabled = true,
                                fingerprint = "chrome"
                            },
                            reality = new
                            {
                                enabled = !string.IsNullOrEmpty(pbk),
                                public_key = pbk,
                                short_id = sid
                            }
                        }
                    },
                    new
                    {
                        type = "direct",
                        tag = "direct"
                    }
                },
                route = new
                {
                    rules = routeRules,
                    auto_detect_interface = true
                }
            };

            File.WriteAllText(configFile, JsonSerializer.Serialize(configObj, new JsonSerializerOptions { WriteIndented = true }));
            _currentConfigFile = configFile;
            return configFile;
        }
        catch
        {
            return null;
        }
    }

    private static object BuildDnsConfig(bool enableAdBlock = true)
    {
        if (enableAdBlock)
        {
            return new
            {
                servers = new object[]
                {
                    new
                    {
                        tag = "dns-filter",
                        type = "udp",
                        server = "127.0.0.1",
                        server_port = 53
                    }
                },
                final = "dns-filter"
            };
        }
        else
        {
            return new
            {
                servers = new object[]
                {
                    new
                    {
                        tag = "dns-direct",
                        type = "udp",
                        server = "1.1.1.1",
                        server_port = 53
                    }
                },
                final = "dns-direct"
            };
        }
    }

    private static object[] BuildRouteRules(bool enableAdBlock, string mainOutbound)
    {
        var rules = new List<object>();

        if (enableAdBlock)
        {
            rules.Add(new
            {
                port = new int[] { 53 },
                action = "hijack-dns"
            });
            rules.Add(new
            {
                port = new int[] { 853 },
                action = "reject"
            });
            rules.Add(new
            {
                network = "udp",
                port = new int[] { 443 },
                action = "reject"
            });
            rules.Add(new
            {
                domain_suffix = new string[]
                {
                    "dns.google",
                    "dns.google.com",
                    "cloudflare-dns.com",
                    "chrome.cloudflare-dns.com",
                    "mozilla.cloudflare-dns.com",
                    "dns.quad9.net",
                    "doh.opendns.com",
                    "dns.nextdns.io",
                    "doh.cleanbrowsing.org",
                    "dns.alidns.com",
                    "doh.pub"
                },
                action = "reject"
            });
            rules.Add(new
            {
                ip_cidr = new string[] { "0.0.0.0/32", "::0/128" },
                action = "reject"
            });
        }

        if (mainOutbound != "direct")
        {
            rules.Add(new
            {
                inbound = new string[] { "mixed-in" },
                action = "route",
                outbound = mainOutbound
            });
        }

        return rules.ToArray();
    }

    private bool StartProcess(string baseDir, string configFile)
    {
        try
        {
            string arch = Environment.Is64BitOperatingSystem ? "x86_64" : "x86";
            string exePath = Path.Combine(baseDir, "engine", arch, "sing-box.exe");

            if (!File.Exists(exePath))
            {
                exePath = Path.Combine(AppDomain.CurrentDomain.BaseDirectory, "engine", arch, "sing-box.exe");
            }

            if (!File.Exists(exePath)) return false;

            string workDir = Path.GetDirectoryName(exePath)!;

            var startInfo = new ProcessStartInfo
            {
                FileName = exePath,
                Arguments = $"run -c \"{configFile}\"",
                WorkingDirectory = workDir,
                UseShellExecute = false,
                CreateNoWindow = true,
                RedirectStandardOutput = true,
                RedirectStandardError = true
            };

            _process = new Process { StartInfo = startInfo, EnableRaisingEvents = true };
            _process.OutputDataReceived += (s, e) => { if (!string.IsNullOrEmpty(e.Data)) AppLogger.Info($"[sing-box] {e.Data}"); };
            _process.ErrorDataReceived += (s, e) => { if (!string.IsNullOrEmpty(e.Data)) AppLogger.Warn($"[sing-box] {e.Data}"); };
            _process.Start();
            _process.BeginOutputReadLine();
            _process.BeginErrorReadLine();
            AppLogger.Info($"[sing-box] Process started with config: {configFile}");
            if (_process == null || _process.HasExited) return false;

            Thread.Sleep(300);
            return !_process.HasExited;
        }
        catch
        {
            return false;
        }
    }

    public void Stop()
    {
        try
        {
            ProxyManager.DisableProxy();

            if (_process != null)
            {
                if (!_process.HasExited)
                {
                    _process.Kill(true);
                    _process.WaitForExit(1000);
                }
                _process.Dispose();
                _process = null;
            }

            foreach (var proc in Process.GetProcessesByName("sing-box"))
            {
                try { proc.Kill(); } catch { }
            }

            if (!string.IsNullOrEmpty(_currentConfigFile) && File.Exists(_currentConfigFile))
            {
                try { File.Delete(_currentConfigFile); } catch { }
                _currentConfigFile = null;
            }
        }
        catch { }
    }
}
